### PR TITLE
[#634] Support defining sinceDate, untilDate and period in report config

### LINKED ISSUE #634: Allow sinceDate and untilDate to be configured in config files

### PR DESCRIPTION
**Pull request by dcshzj**
Fixes #634.
Part of #1360.

```
The new report-config.json allows for the configuration of the
report generation. However, only the report title can be
configured at this time.

Let's expand the functionality of the report config file by
allowing sinceDate, untilDate and period to be defined as well.
If the CLI arguments were to contain the same settings as those in
the report config, then the CLI arguments will take precedence.
```

## Summary of changes
- The `report-config.json` file now supports setting `sinceDate`, `untilDate` and `period`. The equivalent CLI arguments are `--since`, `--until` and `--period` respectively. (this PR)
- The logic for getting the ReportConfiguration object has been moved to ArgsParser. (in PR #1534)
- DateTime functionality that previously existed in ArgsParser has now been moved to TimeUtil to reduce code duplication. (in PR #1535)

## :warning: Requires
- [x] PR #1534 to be merged first
- [x] PR #1535 to be merged first

### CODE REVIEW

**Comment by Tejas2805** [comment]
> The logic for getting the ReportConfiguration object has been moved to ArgsParser.
> DateTime functionality that previously existed in ArgsParser has now been moved to TimeUtil to reduce code duplication.

@dcshzj can you create smaller PRs for these changes. And then after merging those PRs we can go ahead with merging this one. Always remember, smaller neat PRs are preferred as a single PR should bring about only a single change in the repository. This helps in keeping track and reviewing the PR more efficiently.

**Comment by dcshzj** [comment]
> @dcshzj can you create smaller PRs for these changes. And then after merging those PRs we can go ahead with merging this one. Always remember, smaller neat PRs are preferred as a single PR should bring about only a single change in the repository. This helps in keeping track and reviewing the PR more efficiently.

@Tejas2805 I have split them to #1534 and #1535. I left this PR intact and will update if those PRs are merged.

**Comment by dcshzj** [comment]
@Tejas2805 The requisite PRs have been merged.

File: docs/ug/configFiles.md
Comment (Tejas2805) [inline docs/ug/configFiles.md:106]: What's the reason for ignoring all 3? Why not just take since and until in this case, assuming nothing provided in CLI arguments. Don't do changes as of now, just explain the intuition behind this logic for now.

File: src/main/java/reposense/parser/ArgsParser.java
Comment (Tejas2805) [inline src/main/java/reposense/parser/ArgsParser.java:321]: Find the whole logic above a bit cumbersome. There is a better way to avoid a few of the checks by writing the if conditions in a clever manner. I will leave that to you to figure out.

File: docs/ug/configFiles.md
Comment (Tejas2805) [inline docs/ug/configFiles.md:108]: What if all 3 are provided in the report config itself and nothing is provided as CLI arguments?

**Review by Tejas2805 — CHANGES_REQUESTED** [review]
Left few comments for implementation and logic. Will review tests once this is sorted.

File: docs/ug/configFiles.md
Comment (dcshzj) [reply docs/ug/configFiles.md:106]: I followed the logic originally present for CLI arguments, whereby if all 3 arguments are provided we will throw an exception. In this case, because it is in report config, I chose to ignore instead of throwing an exception.

File: docs/ug/configFiles.md
Comment (dcshzj) [reply docs/ug/configFiles.md:108]: Similar to the first part, all 3 will be ignored, as if the user tried to specify all 3 in the CLI arguments instead of the report config.

File: src/main/java/reposense/parser/ArgsParser.java
Comment (dcshzj) [reply src/main/java/reposense/parser/ArgsParser.java:321]: I thought of the following, not sure if that is what you are looking for?

```java
tempPeriod = Optional.ofNullable(cliPeriod.orElse(reportConfig.getPeriod().orElse(null)));
tempSinceDate = Optional.ofNullable(cliSinceDate.orElse(reportConfig.getSinceDate().orElse(null)));
tempUntilDate = Optional.ofNullable(cliUntilDate.orElse(reportConfig.getUntilDate().orElse(null)));
```

**Comment by dcshzj** [comment]
@Tejas2805 Left some comments on your input

**Comment by dcshzj** [comment]
@Tejas2805 Any comments?

**Comment by dcshzj** [comment]
@Tejas2805 Can I please have your comments on this?

File: src/main/java/reposense/parser/ArgsParser.java
Comment (Tejas2805) [reply src/main/java/reposense/parser/ArgsParser.java:321]: I meant for the whole page in general. Not just the last three lines. Cause I can see you see checking if arguments are present or not 2/3 times. So I didn't like that part of the logic. Can try to make sure that you check these only once and make it work.

**Review by Tejas2805 — COMMENTED** [review]
Sorry for the late reply. Left a comment.

File: src/main/java/reposense/parser/ArgsParser.java
Comment (dcshzj) [reply src/main/java/reposense/parser/ArgsParser.java:321]: Sorry @Tejas2805 I still can't figure out how to better write this part of the code. My train of thought for this is:

1. Check if all 3 arguments are provided (either from report config, from CLI, or from both). If yes, we ignore all of them and take only the CLI arguments.
2. For each of the 3 arguments, if it is given in both the CLI and in the report config, we give the user a warning.
3. Afterwards, we take only the CLI arguments.
4. We then process the actual since and until dates, based on the processed information. If both since and until dates are provided (which only happens in CLI), the period is ignored without warning.

I feel that they are all checking different types of conditions, so I can't wrap my head around how to better craft this section.

**Comment by dcshzj** [comment]
Closing this 4-month old PR.

**Pull request closed without merging**
